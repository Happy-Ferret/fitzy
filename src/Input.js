import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Input Button Example
 */
class Input extends Component {
    static propTypes = {
        value: PropTypes.string
    };

    render() {
        return (
            <div>
                <label>Text</label>
                <input type="text" className="lol" value={this.props.value} />
            </div>
        );
    }
}

export default Input;
