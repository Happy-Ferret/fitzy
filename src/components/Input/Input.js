import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Input Button Example
 */
class Input extends Component {
    render() {
        return (
            <div>
                <label>Text</label>
                <input type="text" className="lol" value={this.props.value} />
            </div>
        );
    }
}

Input.propTypes = {
    value: PropTypes.string
};

export default Input;
