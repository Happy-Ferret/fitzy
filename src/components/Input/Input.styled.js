import styled from 'styled-components';

const FormField = styled.label`
  position: relative;
  display: block;
  margin-bottom: 1.5rem;

  /* &:focus-within {
    label {
      top: 0;
    }
  } */
`;

const FormError = styled.span`
  display: block;
  color: red;
`;

const FitzyInput = styled.input`
  display: block;
  position: relative;
  padding: 24px 12px 12px;
  font-size: inherit;
  font-weight: normal;
  font-family: inherit;
  border-radius: 3px;
  border: 1px solid #ccc;
  width: 100%;
  background-color: #fff;
  transition: border 0.3s ease-in-out;
  margin: 0 0 5px 0;

  ~ label {
    top: ${({ value }) => (value ? '0' : '18px')};
  }
`;

const FitzyLabel = styled.label`
  display: inline-block;
  max-width: 100%;
  position: absolute;
  left: 12px;
  z-index: 1;
  transition: all 0.3s ease;
`;

/** @component */
export { FitzyInput, FormField, FitzyLabel, FormError };
