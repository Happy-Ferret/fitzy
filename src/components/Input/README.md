Basic Input :

```jsx
initialState = { value: '' };

<Input
  placeholder="Enter your name"
  name="name1"
  value={state.value}
  onChange={(_, v) => setState({ value: v })}
/>;
```

Dark :

```jsx
initialState = {
  username: '',
  password: '',
  name: 'John Doe',
  email: 'johndoe.com'
};

<div className="flex dark column">
  <Input
    label="Username"
    placeholder="Enter your name"
    name="username"
    value={state.username}
    onChange={(_, v) => setState({ username: v })}
    required
  />
  <Input
    label="Password"
    placeholder="Enter your password"
    name="password"
    value={state.password}
    onChange={(_, v) => setState({ password: v })}
    required
  />
  <Input
    label="Email"
    placeholder="Enter your email"
    name="email2"
    value={state.email}
    onChange={(_, v) => setState({ email: v })}
    warning={<span>Invalid Email</span>}
    required
  />
  <Input
    label="goodbye"
    placeholder="Enter your name"
    name="name"
    value={state.name}
    onChange={(_, v) => setState({ name: v })}
    required
  />
</div>;
```

Input with warning :

```jsx
initialState = { value: '' };

<Input
  label="Name"
  value={state.value}
  onChange={(_, v) => setState({ value: v })}
  warning={<span>Some Warning</span>}
/>;
```

Input with label :

```jsx
initialState = { value: '' };

<Input
  label="Name"
  name="name2"
  value={state.value}
  onChange={(_, v) => setState({ value: v })}
  autoComplete="username"
  required
/>;
```
