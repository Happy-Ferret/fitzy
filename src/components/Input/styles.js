import styled from 'styled-components';

const FormField = styled.label`
  position: relative;
  display: block;
  margin-bottom: 1.5rem;
`;

const FormError = styled.span`
  display: block;
  color: red;
`;

const FitzyInput = styled.input`
  display: block;
  position: relative;
  padding: 12px;
  font-size: inherit;
  font-weight: normal;
  border-radius: 3px;
  border: 1px solid #ccc;
  font-family: inherit;
  width: 100%;
  background-color: #fff;
  transition: all 0.3s ease-in-out;
  margin: 0 0 5px 0;
`;

const FitzyLabel = styled.label`
  display: inline-block;
  margin-bottom: 5px;
  font-weight: bold;
  max-width: 100%;
`;

export { FitzyInput, FormField, FitzyLabel, FormError };
