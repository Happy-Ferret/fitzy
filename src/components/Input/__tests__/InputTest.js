import React from 'react';
import { shallow } from 'enzyme';

import Input from '../Input';

describe('Input', () => {
  let wrapper;
  let input;

  const onChangeSpy = jest.fn().mockName('onChange');
  const onBlurSpy = jest.fn().mockName('onBlur');

  const defaultProps = {
    type: 'text',
    value: 'test',
    name: 'name',
    placeholder: 'Enter Name',
    onChange: onChangeSpy,
    onBlur: onBlurSpy
  };

  const funkyProps = {
    ...defaultProps,
    spellCheck: false,
    autoComplete: 'username',
    autoFocus: true
  };

  beforeEach(() => {
    wrapper = shallow(<Input type="text" {...defaultProps} />);
    input = wrapper.find('input');
  });

  afterEach(() => {
    onChangeSpy.mockReset();
  });

  it('should render Input field with type text', () => {
    expect(wrapper.childAt(0).type()).toEqual('input');
    expect(input.props().type).toEqual('text');
  });

  it('should handle default props', () => {
    expect(input.props().placeholder).toEqual('Enter Name');
    expect(input.props().disabled).toEqual(false);
  });

  it('should set id based on name if no id entered', () => {
    expect(input.props().id).toEqual(defaultProps.name);
  });

  it('should call onChange when text entered', () => {
    expect(onChangeSpy.mock.calls).toHaveLength(0);
    input.simulate('change', { target: { value: 'new' } });
    expect(onChangeSpy.mock.calls).toHaveLength(1);
    expect(onChangeSpy.mock.calls[0][1]).toBe('new');
  });

  it('should call onBlur when leaving input', () => {
    expect(onBlurSpy.mock.calls).toHaveLength(0);
    input.simulate('blur', { target: { value: 'new' } });
    expect(onBlurSpy.mock.calls).toHaveLength(1);
  });

  describe('with Label', () => {
    beforeEach(() => wrapper.setProps({ label: 'Name:' }));

    it('should render label when set', () => {
      expect(wrapper.childAt(0).type()).toEqual('label');
    });

    it('should set text and for correctly', () => {
      expect(wrapper.childAt(0).text()).toEqual('Name:');
      expect(wrapper.childAt(0).props().htmlFor).toEqual('name');
    });
  });

  describe('with Hint', () => {
    let hint;

    beforeEach(() => {
      wrapper.setProps({ hint: 'Only Unique' });
      hint = wrapper.find('p');
    });

    it('should render hint when set', () => {
      expect(hint).toHaveLength(1);
      expect(hint.containsMatchingElement(<p>Only Unique</p>)).toEqual(true);
    });

    it('should set text correctly', () => {
      expect(hint.text()).toEqual('Only Unique');
    });
  });

  describe('with Checkbox', () => {
    beforeEach(() => {
      wrapper.setProps({ type: 'checkbox' });
      input = wrapper.find('input');
    });

    it('should render type of checkbox', () => {
      expect(input.props().type).toEqual('checkbox');
    });

    it('should call onChange with checked value', () => {
      expect(onChangeSpy).not.toBeCalled();
      const event = { target: { checked: false } };
      input.simulate('change', event);
      expect(onChangeSpy).toBeCalled();
      expect(onChangeSpy).toBeCalledWith(event, false);
    });
  });

  describe('with TextArea', () => {
    beforeEach(() => {
      wrapper.setProps({ type: 'textarea' });
      input = wrapper.find('textarea');
    });

    it('should render type of checkbox', () => {
      expect(input.props().type).toEqual('textarea');
      expect(input.is('textarea')).toEqual(true);
    });
  });

  describe('with Funky Props', () => {
    it('should display all extra funky props', () => {
      wrapper.setProps({ ...funkyProps });
      const funkyInput = wrapper.find('input');
      expect(Object.keys(funkyInput.props())).toEqual(
        expect.arrayContaining(Object.keys(funkyProps))
      );
    });
  });
});
