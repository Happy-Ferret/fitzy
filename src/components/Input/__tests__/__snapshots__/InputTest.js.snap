// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot 1`] = `
.c0 {
  position: relative;
  display: block;
  margin-bottom: 1.5rem;
}

.c1 {
  display: block;
  position: relative;
  padding: 24px 12px 12px;
  font-size: inherit;
  font-weight: normal;
  font-family: inherit;
  border-radius: 3px;
  border: 1px solid #ccc;
  width: 100%;
  background-color: #fff;
  -webkit-transition: border 0.3s ease-in-out;
  transition: border 0.3s ease-in-out;
  margin: 0 0 5px 0;
}

.c1 ~ label {
  top: 0;
}

<label
  className="c0"
  role="presentation"
>
  <input
    autoComplete={null}
    autoFocus={null}
    className="c1"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Enter Name"
    required={false}
    spellCheck={null}
    type="text"
    value="test"
  />
</label>
`;
